profile/src/main/java/com/scottlogic/deg/profile/reader/JsonProfileReader.java

```diff
@@ -38,115 +43,108 @@
-
-        if (profileDto.fields == null) {
-            throw new InvalidProfileException("Profile is invalid: 'fields' have not been defined.");
-        }
-        if (profileDto.rules == null) {
-            throw new InvalidProfileException("Profile is invalid: 'rules' have not been defined.");
-        }
-
-        List<Field> inMapFields = getInMapConstraints(profileDto).stream()
-            .map(file -> new Field(file, SpecificFieldType.from("integer").getFieldType(),false,null, true)
-            ).collect(Collectors.toList());
+    public Profile read(String profileJson) throws IOException
+    {
+        ProfileDTO profileDTO = new ProfileSerialiser().deserialise(profileJson);
+        if (profileDTO.fields == null) throw new InvalidProfileException("Profile is invalid: 'fields' have not been defined.");
+        if (profileDTO.rules == null) throw new InvalidProfileException("Profile is invalid: 'rules' have not been defined.");
 
+        List<Field> fields = profileDTO.fields.stream()
+                .map(fieldDTO -> new Field(fieldDTO.name, fieldDTO.type.getFieldType(), fieldDTO.unique, fieldDTO.formatting, false))
+                .collect(Collectors.toList());
 
-        List<Field> fields = profileDto.fields.stream()
-            .map(fDto -> new Field(fDto.name, fDto.type.getFieldType(), fDto.unique, fDto.formatting,false))
-            .collect(Collectors.toList());
+        List<Field> inMapFields = profileDTO.rules.stream()
+                .flatMap(ruleDTO -> ruleDTO.constraints.stream())
+                .flatMap(constraintDTO -> getInMapConstraints(profileDTO).stream())
+                .distinct()
+                .map(file -> new Field(file, FieldType.NUMERIC, false, null, true)
+                ).collect(Collectors.toList());
 
         fields.addAll(inMapFields);
-
         ProfileFields profileFields = new ProfileFields(fields);
 
-        Collection<Rule> rules = profileDto.rules.stream().map(
-            r -> {
-                RuleInformation constraintRule = new RuleInformation(r.rule);
-                return new Rule(constraintRule, mainConstraintReader.getSubConstraints(profileFields, r.constraints));
-            }).collect(Collectors.toList());
-
-
-        // add nullable
-        Collection<Constraint> nullableRules = profileDto.fields.stream()
-            .filter(fieldDTO -> !fieldDTO.nullable)
-            .map(fieldDTO -> create(AtomicConstraintType.IS_NULL, profileFields.getByName(fieldDTO.name), null).negate())
-            .collect(Collectors.toList());
+        Collection<Rule> rules = profileDTO.rules.stream()
+                .map(r -> new Rule(new RuleInformation(r.rule), constraintReader.read(r.constraints, profileFields)))
+                .collect(Collectors.toList());
 
-        if (nullableRules.size() > 0) {
-            rules.add(new Rule(new RuleInformation("nullable-rules"), nullableRules));
-        }
+        Collection<Constraint> nullableConstraints = profileDTO.fields.stream()
```

Comment:
I think that this collection actually holds not null constraints.

Perhaps rename it to notNullableConstraints or something similar

Reply:
This has been renamed in #1461